apps/aechannel/src/aesc_fsm_min_depth_watcher.erl

```diff
@@ -104,173 +293,811 @@ terminate(_Reason, _St) ->
+            %% If the channel state is undefined, assume that the channel has
+            %% been deleted. A corner case is that it re-appears after e.g. a
+            %% fork switch, but with a suitable min_depth value, it should be ok
+            %% to give up after TopHeight + MinDepth
+            {TopHeight, C1} = top_height(C),
+            Min = maps:get(min_depth, R, 0),
+            {R#{ locked_until    => TopHeight
+               , check_at_height => TopHeight + Min }, C1}
+    end;
+check_req(#{mode := unlock} = R, #st{chan_id = ChId, chan_vsn = Vsn} = St, C) ->
+    if Vsn =/= undefined ->
+            lager:debug("checking", []),
+            case get_basic_ch_status_(ChId, C) of
+                {#{locked_until := 0} = Ch, C1} ->
+                    lager:debug("locked_until = 0", []),
+                    report_channel_unlocked(R, Ch, St, C1),
+                    {done, C1};
+                {#{locked_until := LockedUntil} = Ch, C1} ->
+                    {Height, C2} = top_height(C1),
+                    lager:debug("LockedUntil = ~p, Height = ~p", [LockedUntil, Height]),
+                    case LockedUntil < Height of
+                        true  ->
+                            lager:debug("locked_until expired", []),
+                            report_channel_unlocked(R, Ch, St, C2),
+                            {done, C2};
+                        false ->
+                            lager:debug("still locked", []),
+                            {R#{check_at_height => LockedUntil + 1}, C2}
+                    end;
+                {undefined, C1} ->
+                    lager:debug("couldn't find channel", []),
+                    {done, C1}
+            end;
+       true ->
+            lager:debug("Vsn = undefined", []),
+            {R, C}
+    end;
+check_req(#{mode := watch} = R, #st{chan_id = ChId} = St,
+          #{scenario := Scenario, top_hash := Hash } = C)
+  when ?IS_SCENARIO(Scenario) ->
+    lager:debug("Scenario = ~p", [Scenario]),
+    case Scenario of
+        {has_tx, {{TxHash,_}, #{ type    := TxType }}} ->
+            {Ch, C1} = get_channel(ChId, Hash, C),
+            {#{ height := H }, C2} = top_info(Hash, C1),
+            case Ch of
+                undefined when TxType == channel_close_mutual_tx;
+                               TxType == channel_settle_tx ->
+                    report_closed_on_chain(#{ tx_type => TxType
+                                            , tx_hash => TxHash
+                                            , block_hash => Hash
+                                            , height  => H }, R, St, C2);
+                _ ->
+                    {Status, C3} = channel_status(ChId, C2),
+                    lager:debug("Status = ~p", [Status]),
+                    watch_for_change_in_ch_status(Status, H, R, St, C3)
+            end;
+        fork_switch ->
+            watch_for_channel_change(R, St, C);
+        _ ->
```

Comment:
I see that scenario `top` gets this request `watch` discarded.

Reply:
Well, the `watch` request will remain in play, but will not perform a check in a `top` scenario, which triggers only when a new request object is introduced.

Reply:
Ah! the watch request is not discarded but placed in `requests` field of state. Thanks.